src/Conversion/ONNXToKrnl/Tensor/ReverseSequence.cpp

```diff
@@ -0,0 +1,133 @@
+  ONNXReverseSequenceOpLowering(MLIRContext *ctx)
+      : ConversionPattern(
+            mlir::ONNXReverseSequenceOp::getOperationName(), 1, ctx) {}
+
+  LogicalResult matchAndRewrite(Operation *op, ArrayRef<Value> operands,
+      ConversionPatternRewriter &rewriter) const final {
+    ONNXReverseSequenceOpAdaptor operandAdaptor(operands);
+    ONNXReverseSequenceOp reverseSequenceOp =
+        llvm::cast<ONNXReverseSequenceOp>(op);
+    auto loc = op->getLoc();
+
+    ONNXReverseSequenceOpShapeHelper shapeHelper(&reverseSequenceOp, rewriter,
+        getDenseElementAttributeFromKrnlValue,
+        loadDenseElementArrayValueAtIndex);
+    auto shapecomputed = shapeHelper.Compute(operandAdaptor);
+    assert(succeeded(shapecomputed));
+    // Scope for krnl ops
+    IndexExprScope outerScope(rewriter, shapeHelper.scope);
+    KrnlBuilder createKrnl(rewriter, loc);
+
+    // Insert an allocation and deallocation for the output of this operation.
+    MemRefType outputMemRefType = convertToMemRefType(*op->result_type_begin());
+    Value alloc = insertAllocAndDeallocSimple(
+        rewriter, op, outputMemRefType, loc, shapeHelper.dimsForOutput(0));
+
+    // Save axis and rank info.
+    int64_t batchAxis = reverseSequenceOp.batch_axis();
+    int64_t timeAxis = reverseSequenceOp.time_axis();
+
+    MemRefBoundsIndexCapture dataBounds(operandAdaptor.input());
+    int64_t outputRank = shapeHelper.dimsForOutput(0).size();
+
+    /*
+      The semantic of Reversequence can be expressed in loop as:
+
+      for (vector of dim I on the shape of output tensor) {
+        vector of dim Iinput = I;
+        Iinput[time_axis()]  = I[time_axis()] < seqence_lens[I[batch_axis()]]?
+              sequence_lens[I[batch_axis()]]-I[time_axis()]-1:I[time_axis()];
+        output[I] = input[Iinput]
+      }
+
+      Obviously, since the conditional check is on time_axis() loop variable,
+      the check can be eliminated with loop splitting as long as the batch_axis
+      loop is outside.
+
+      I is vector of dim on the shape of output tensor;
+      for ( I except I[batch_axis()]) {
+        // Reverse
+        for (I[time_axis() :  0 to sequence_lens[I[batch_axis]]) {
+          Iinput = I;
+          Iinput[time_axis()] = sequence_lens[I[batch_axis()]]-I[time_axis()]-1
+          output[I] = input[Iinput];
+        }
+        // Copy
+        for (I[time_axis() from sequence_lens[I[batch_axis]] to end) {
+          output[I] = input[I];
+        }
+      }
+
+      This transformation should improve performance on this loop nest itself,
+      but may hinder loop fusion with other loop nestes.
+      Also further loop fission or loop interchanging can be applied here.
+      I chose the simple loop structure
+      and believe optimization should be left to compiler, at least for
+      non-critical ops.
+    */
+
+    // Define loops and iteration trip counts (equivalent to size of output)
```

Comment:
While there is nothing wrong with using the old interface, the new lambda based version is so much easier to read, I would strongly suggest to use it from now on. Ideally we would also update the old versions, but if we could generate the new ones using the new scheme, that would be great. If you believe something is preventing you, I would be happy to add the necessary support.

Reply:
Any documentation for the new lambda approach?